sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts

```diff
@@ -0,0 +1,155 @@
+import { ServiceBusClient, delay, ReceivedMessageWithLock } from "@azure/service-bus";
+import * as dotenv from "dotenv";
+import { env } from "process";
+import { AbortController, AbortSignalLike } from "@azure/abort-controller";
+
+dotenv.config();
+
+const serviceBusConnectionString =
+  env["SERVICEBUS_CONNECTION_STRING"] || "<service bus connection string not in environment>";
+const queueName = env["QUEUE_NAME_WITH_SESSIONS"] || "<queue name not in environment>";
+const maxSessionsToProcessSimultaneously = 8;
+const sessionIdleTimeoutMs = 3 * 1000;
+const delayWhenNoSessionsAvailableMs = 5 * 1000;
+
+interface SessionMessageHandlers {
+  processInitialize(sessionId: string);
+  processMessage(msg: ReceivedMessageWithLock);
+  processError(err: Error, sessionId: string);
+  processClose(reason: "error" | "idle_timeout", sessionId: string);
+}
+
+function _createIdleTimer(
+  timeoutMs: number
+): { refresh(): void; expirationPromise: Promise<void> } {
+  let resolveExpirationPromise;
+
+  const expirationPromise = new Promise<void>((res, _rej) => {
+    resolveExpirationPromise = res;
+  });
+
+  let timer = setTimeout(resolveExpirationPromise, timeoutMs);
+
+  return {
+    refresh() {
+      clearTimeout(timer);
+      timer = setTimeout(resolveExpirationPromise, timeoutMs);
+    },
+    expirationPromise
+  };
+}
+
+async function _noSessionsAvailable(err: Error) {
+  // TODO: this error should be programmatically identifable.
+  return err.message.search(/Received an incorrect sessionId 'undefined'/) !== -1;
+}
+
+async function _processNextSession(
+  serviceBusClient: ServiceBusClient,
+  handlers: SessionMessageHandlers,
+  abortSignal: AbortSignalLike
+): Promise<void> {
+  const sessionReceiver = serviceBusClient.createSessionReceiver(queueName, "peekLock", {
+    autoRenewLockDurationInMs: sessionIdleTimeoutMs
+  });
+
+  try {
+    // TODO: I just picked a call that didn't grab a message here, just to ensure a session could
+    // actually be obtained. This goes away when we have `await createSessionReceiver()`.
+    await sessionReceiver.getState();
+  } catch (err) {
+    // TODO: when this happens do I still need to close the session receiver? (again, temporary concern)
+    if (_noSessionsAvailable(err)) {
+      console.log(`INFO: no available sessions, sleeping for ${delayWhenNoSessionsAvailableMs}`);
+      await delay(delayWhenNoSessionsAvailableMs);
+      return;
+    }
+
+    throw err;
+  }
+
+  await handlers.processInitialize(sessionReceiver.sessionId);
+
+  const idleTimer = _createIdleTimer(sessionIdleTimeoutMs);
+  let threwError = false;
+
+  sessionReceiver.subscribe(
+    {
+      async processMessage(msg) {
+        idleTimer.refresh();
+        await handlers.processMessage(msg);
+      },
+      async processError(err) {
+        threwError = true;
+        await handlers.processError(err, sessionReceiver.sessionId);
```

Comment:
All errors other than the ones stemming from running `processMessage` should be considered as fatal and should result in us moving to the next session.

Reply:
I need to check - do we funnel errors thrown by the user in processMessage here too? Do we provide an 'isFatal' flag or something on our internal errors?

Reply:
Yes, we do funnel the errors thrown by the user to `processError`.
From what I recall, an error can be an instance of the custom class `MessagingError` (in which case there is a retryable property on it) or it could be on of the common error class like `Error` and `TypeError`.

So, all retryable `MessagingError` errors should be treated as non fatal and allow moving to the next session

Reply:
But, on the other hand `SessionLockLostError` is not retryable, but I would like to keep the loop going for this error...

Reply:
I simplified it some more to make it so that an error or the idle timeout cause us to close the session receiver rather than try to dive into error handling here.

Reply:
CC: @ramya-rao-a